rust/core-lib/src/event_context.rs

```diff
@@ -494,6 +495,29 @@ impl<'a> EventContext<'a> {
         self.editor.borrow_mut().dec_revs_in_flight();
     }
 
+    /// Returns the text to be saved, appending a newline if necessary.
+    pub(crate) fn text_for_save(&mut self) -> Rope {
+        let editor = self.editor.borrow();
+        let mut rope = editor.get_buffer().clone();
+        let rope_len = rope.len();
+        let last_char = rope.byte_at(rope.len() - 1) as char;
+
+        // Only add a new line if the user actually wrote something,
+        // there's no newline at the end of the file
+        // and the config option for it is true
+        if rope_len > 0
+            // If parse_chunk fails just play it safe and return Some so we don't add a superfluous newline
+            && LineEnding::parse_chunk(&last_char.to_string()).unwrap_or(Some(LineEnding::Lf)).is_none()
```

Comment:
this is ~pretty~ very confusing.

Can this ever fail? I... think it can't?

I think that the previous version, with just checking for `\n` at EOF is fine. However, if we want to do this _actually_ correctly, I would suggest the following approach:

```rust
let mut cursor = Cursor::new(&rope, rope.len());
let newline_at_eof = match cursor.get_leaf() {
    Some((last_chunk, _)) => last_chunk.ends_with(&self.config.line_ending),
    // there's no last chunk, which means rope is empty, but we checked that earlier?  ¯\_(ツ)_/¯
   // this is only true of course if we _do_ check for is_empty...
    None => unreachable!(),
};

if self.config.save_with_newline && newline_at_eof { ... }
```

Does that maybe make sense? We do the cursor business because it lets us efficiently get the actual string value for the last chunk of the rope.

Reply:
> I think that the previous version, with just checking for \n at EOF is fine

Hum, if we already have LineEnding stuff we might as well use that to make sure we have everything in one place

Cursor still sort of is magic to me, will look into that tomorrow

Reply:
Alright, I just use config.line_ending now, since it's not really relevant what kind of line ending is used here. Does this work if the user sets line ending x in the config and xi autodetects a different line ending in the file though?